### PR TITLE
good enough wide string support

### PR DESCRIPTION
**Pull request by Lokathor**
Various TODOs:

* [x] Tag everything as `#[inline]` probably since a lot of this is non-generic (generics never need to have the `#[inline]` attribute)
* [x] Docs on everything
* [x] AsRef?
* [x] Borrow?
* [x] BorrowMut?
* [x] Default
* [x] Indexing
* [x] Slicing
* [x] Something so that we can go from wide back to normal (probably with String output)
* [x] Make WZStr deref into WStr and then remove most inherent methods from WZStr?
* [x] ~~should `AsWide` be bound by `Deref<WStr>`?~~
* [x] Review: why do we have `IntoWide` and `IntoWideNull`? Can't we use From/Into?
* [x] Do we even want `AsWide` and `AsWideNull` instead of just having `Deref<WStr>` and `Deref<WZStr>`?
* [x] Method to try and "promote" a `WStr` to a `WZStr` (and it fails if you didn't start off null terminated)
* [x] Message box should take `IntoWideNull`
* [x] SIMD enabled equality checking? Because why not.
* [x] PartialEq / Eq between WStr and WZStr
* [x] FromIter<u16>
* [x] any string manipulation routines?
* [x] Review `String` and `str` for capability coverage
* [x] Tests for all sorts of stuff because we're probably less smart than we think we are.

### CODE REVIEW

File: src/wide_string.rs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+pub trait AsWide {
+  fn as_wide(&self) -> &WStr;
+  fn as_wide_ptr(&self) -> *const u16 {
+    self.as_wide().as_ptr()
+  }
+  fn len(&self) -> usize {
+    self.as_wide().len()
+  }
+}
+impl<T: ?Sized> AsWide for &T
```
Comment (retep998) [inline src/wide_string.rs]: Maybe just `<T: ?Sized + AsWide>`?

Comment (Lokathor) [reply src/wide_string.rs]:
```rust
impl<T: ?Sized + AsWide> AsWide for &T
where
  T: AsWide,
{
  fn as_wide(&self) -> &WStr {
    (*self).as_wide()
  }
}
```

File: src/wide_string.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+}
+impl<T> IntoWideNull for T
+where
+  T: AsWideNull,
+{
+  type WideNull = T;
+  fn into_wide_null(self) -> Self::WideNull {
+    self
+  }
+}
```
Comment (retep998) [inline src/wide_string.rs]:
```Rust

impl IntoWideNull for WString {
    type WideNull = WZString;
    fn into_wide_null(self) -> Self::WideNull {
        self.make_null_terminated()
    }
}
```

Comment (Lokathor) [reply src/wide_string.rs]: got it

File: src/wide_string.rs (hunk 1 above)
Comment (retep998) [reply src/wide_string.rs]: So that you don't need the where bound at all taking up all those lines.

File: src/wide_string.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        self.0.push(0);
+        Some(WZString(self.0))
+      }
+    }
+  }
+}
+impl Deref for WString {
+  type Target = WStr;
+  fn deref(&self) -> &WStr {
+    unsafe { *(self.0.as_ptr() as *const _) }
```
Comment (retep998) [inline src/wide_string.rs]: `unsafe { &*(&*self.0 as *const [u16] as *const WStr) }`

File: src/wide_string.rs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+  }
+}
+
+/// The owned version of a wide, null terminated string
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct WZString(Vec<u16>);
+impl Deref for WZString {
+  type Target = WZStr;
+  fn deref(&self) -> &WZStr {
+    unsafe { *(self.0.as_ptr() as *const _) }
```
Comment (retep998) [inline src/wide_string.rs]: `unsafe { &*(&*self.0 as *const [u16] as *const WZStr) }`

File: src/wide_string.rs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+  }
+}
+impl AsWide for WZString {
+  fn as_wide(&self) -> &WStr {
+    (**self).as_wide()
+  }
+}
+impl AsWide for WZStr {
+  fn as_wide(&self) -> &WStr {
+    unsafe { *(self as *const _ as *const _) }
```
Comment (retep998) [inline src/wide_string.rs]: `unsafe { &*(self as *const WZStr as *const WStr) }`

File: src/wide_string.rs (hunk 1 above)
Comment (Lokathor) [reply src/wide_string.rs]: oh i see!

File: src/wide_string.rs [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+  fn as_wide(&self) -> &WStr;
+  fn as_wide_ptr(&self) -> *const u16 {
+    self.as_wide().as_ptr()
+  }
+  fn len(&self) -> usize {
+    self.as_wide().len()
+  }
+}
+impl<T: ?Sized + AsWide> AsWide for &T
+{
```
Comment (retep998) [inline src/wide_string.rs]: WHAT IS THIS NEWLINE DOING HERE?

File: src/wide_string.rs [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+  fn as_wide(&self) -> &WStr {
+    unsafe { &*(self as *const WZStr as *const WStr) }
+  }
+}
+
+/// Trait for types which can be viewed as wide null strings.
+pub trait AsWideNull: AsWide {
+  fn as_wide_null(&self) -> &WZStr;
+}
+impl<T: ?Sized  +AsWideNull> AsWideNull for &T
```
Comment (retep998) [inline src/wide_string.rs]: YOUR SPACING IS ATROCIOUS

Comment (Lokathor) [reply src/wide_string.rs]: i blame rustfmt for not fixing it

File: src/wide_string.rs [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+// TODO: impls for all std string types here
+
+/// Trait for anything you can convert into a wide null string.
+///
+/// This often involves an allocation.
+pub trait IntoWideNull: IntoWide {
+  type WideNull: AsWideNull;
+  fn into_wide_null(self) -> Self::WideNull;
+}
+impl<T> IntoWideNull for T
```
Comment (retep998) [inline src/wide_string.rs]: This can also have the bound changed into the shorter `<T: AsWideNull>`

File: src/wide_string.rs [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+  fn as_wide(&self) -> &WStr {
+    unsafe { &*(self as *const WZStr as *const WStr) }
+  }
+}
+
+/// Trait for types which can be viewed as wide null strings.
+pub trait AsWideNull: AsWide {
+  fn as_wide_null(&self) -> &WZStr;
+}
+impl<T: ?Sized + AsWideNull> AsWideNull for &T {
```
Comment (retep998) [inline src/wide_string.rs]: Personally I'd reorder it to be `AsWideNull + ?Sized`.

File: src/bin/thorium_demo.rs [hunk 10]
```diff
@@ -1,11 +1,14 @@
...
+};
+//use wio::wide::ToWide;
 
 fn main() {
   message_box(
     core::ptr::null_mut(),
-    "Hello Message".to_wide_null(),
-    Some("Hello Title".to_wide_null()),
+    "Hello Message".into_wide_null(),
+    Some("Hello Title".into_wide_null()),
```
Comment (retep998) [inline src/bin/thorium_demo.rs]: If `message_box` accepts `T: IntoWideNull` then you don't need to do any conversions here. Just pass `"Hello Message"` directly!

File: src/message_box.rs [hunk 11]
```diff
@@ -202,28 +203,25 @@ impl MessageBoxTypeBuilder {
 ///
 /// Other error results are also possible, but MSDN won't tell you which ones
 /// exactly because it depends on your Windows version and stuff.
-pub fn message_box(
-  owner: HWND, text: Vec<u16>, opt_title: Option<Vec<u16>>, box_type: MessageBoxType,
+pub fn message_box<M: AsWideNull, T: AsWideNull>(
```
Comment (retep998) [inline src/message_box.rs]: Take `IntoWideNull` as the bound for your generic parameters.

**Comment by Lokathor** [comment]
@retep998 should `AsWide` be bound by `Deref<WStr>`?

**Comment by retep998** [comment]
@Lokathor No, because it isn't necessary. `AsWide` has a method that provides the `WStr` and that's good enough.

**Pull request merged**